File: lib/combo/custom.rb
require 'warning'
Warning.ignore(//, /.*ruby-elf.*/)
require 'elf'
require 'combo/common'
require 'combo/object_splitter'

module Combo
  class Custom
    def self.generate()
      self.new.run()
    end

    def initialize()
      @buffer = StringIO.new
      @dma = []
      @objects = []
      @defines = []
    end

    def run()
      add_split_object("GI_REMAINS_ODOLWA",   :mm, "objects/object_bsmask", [0x0690])
      add_split_object("GI_REMAINS_GOHT",     :mm, "objects/object_bsmask", [0x3ad0])
      add_split_object("GI_REMAINS_GYORG",    :mm, "objects/object_bsmask", [0x1d80])
      add_split_object("GI_REMAINS_TWINMOLD", :mm, "objects/object_bsmask", [0x5020])

      custom_object_files = Dir.glob(File.join(Combo::PATH_BUILD, 'obj/objects/**/*.o'))
      custom_object_files.each do |file|
        add_custom_object(file)
      end
      write_custom()
    end

    def add_split_object(name, game, filename, offsets)
      dma_index = Combo::DATA_FILES[game].index(filename)
      object = nil
      File.open(File.join(Combo::PATH_BUILD, 'roms', "#{game}_decompressed.z64"), "rb") do |f|
        f.seek(Combo::METADATA[game][:dma_addr])
        dma_data = Combo::DmaData.new(f.read(Combo::METADATA[game][:dma_count] * 10))
        entry = dma_data.entry(dma_index)
        f.seek(entry.vstart)
        object = f.read(entry.vlen)
      end
      object_new = ObjectSplitter.run(object, offsets)
      exports = object_new.offsets.map.with_index{|addr, i| [i, addr]}.to_h
      add_object(name, object_new.data, exports)
    end

    def add_custom_object(file)
      File.open(file, 'rb') do |f|
        Elf::File.open(file) do |elf|
          cobject_data = elf.load_section(".cobject.data")
          cobject_name = elf.load_section(".cobject.name")

          f.seek(cobject_data.offset)
          data = f.read(cobject_data.size)

          f.seek(cobject_name.offset)
          name = f.read(cobject_name.size).unpack('Z*').first

          cobject_exports_addr = elf.load_section(".cobject.exports_addr")
          f.seek(cobject_exports_addr.offset)
          exports_addr = f.read(cobject_exports_addr.size).unpack('L>*')

          cobject_exports_name = elf.load_section(".cobject.exports_name")
          f.seek(cobject_exports_name.offset)
          exports_name = f.read(cobject_exports_name.size).split(0.chr)

          exports = {}
          exports_addr.size.times do |i|
            addr = exports_addr[i]
            n = exports_name[i]
            exports[n] = addr
          end

          add_object(name, data, exports)
        end
      end
    end

    def add_object(name, data, exports)
      exports.each do |k, v|
        defname = ["CUSTOM_OBJECT", name, k].join('_')
        add_define(defname, v)
      end
      id = append_vrom_object(data)
      add_define(["CUSTOM_OBJECT_ID", name].join('_'), id)
    end

    def append_vrom_object(data)
      id = @objects.size | 0x2000
      @objects.push(append_vrom_dma(data))
      id
    end

    def append_vrom_dma(data)
      paddr_start = Combo::METADATA[:custom_addr] + @buffer.size
      paddr_end = paddr_start + data.size
      vaddr_start = append_vrom(data)
      vaddr_end = vaddr_start + data.size
      @dma.push([vaddr_start, vaddr_end, paddr_start, 0])
      [vaddr_start, vaddr_end]
    end

    def append_vrom(data)
      off = @buffer.size
      @buffer.write(data)
      return off | 0x08000000
    end

    def write_custom()
      custom_addr_dma = Combo::METADATA[:custom_addr] + @buffer.size
      dma_data = @dma.map { |dma| dma.pack('L>4') }.join('')
      @buffer.write(dma_data)

      custom_addr_objects = Combo::METADATA[:custom_addr] + @buffer.size
      objects_data = @objects.map { |object| object.pack('L>2') }.join('')
      @buffer.write(objects_data)

      add_define("CUSTOM_DMA_ADDR", custom_addr_dma)
      add_define("CUSTOM_DMA_SIZE", @dma.size)
      add_define("CUSTOM_OBJECTS_ADDR", custom_addr_objects)
      add_define("CUSTOM_OBJECTS_SIZE", @objects.size)

      File.open(File.join(Combo::PATH_BUILD, 'include', 'combo', 'custom.h'), "w") do |f|
        f.puts("#ifndef COMBO_CUSTOM_H")
        f.puts("#define COMBO_CUSTOM_H")
        f.puts("")
        f.puts(@defines.join("\n"))
        f.puts("")
        f.puts("#endif")
      end

      File.binwrite(File.join(Combo::PATH_BUILD, 'custom.bin'), @buffer.string)
    end

    def add_define(name, value)
      @defines.push("#define %s 0x%08x" % [name, value])
    end
  end
end

File: lib/combo/custom.ts
import path from 'path';
import fs from 'fs/promises';
import { Buffer } from 'buffer';

import { Game, DATA_FILES, CUSTOM_ADDR } from './config';
import { DmaData } from './dma';
import { splitObject } from './split';
import { align, arrayToIndexMap } from './util';
import { compressFile } from './compress';
import { CodeGen } from './codegen';
import { DecompressedRoms } from './decompress';
import { Monitor } from './monitor';

const FILES_TO_INDEX_OOT = arrayToIndexMap(DATA_FILES.oot);
const FILES_TO_INDEX_MM = arrayToIndexMap(DATA_FILES.mm);
const FILES_TO_INDEX = {
  oot: FILES_TO_INDEX_OOT,
  mm: FILES_TO_INDEX_MM,
};

type CustomEntry = {
  game: Game,
  name: string,
  file: string,
  offsets: number[],
};

const ENTRIES: CustomEntry[] = [
  { game: 'mm',  name: "GI_REMAINS_ODOLWA",     file: "objects/object_bsmask",       offsets: [0x0690] },
  { game: 'mm',  name: "GI_REMAINS_GOHT",       file: "objects/object_bsmask",       offsets: [0x3ad0] },
  { game: 'mm',  name: "GI_REMAINS_GYORG",      file: "objects/object_bsmask",       offsets: [0x1d80] },
  { game: 'mm',  name: "GI_REMAINS_TWINMOLD",   file: "objects/object_bsmask",       offsets: [0x5020] },
  { game: 'oot', name: "GI_MASTER_SWORD",       file: "objects/object_toki_objects", offsets: [0x1bd0] },
];

const getObjectBuffer = async (roms: DecompressedRoms, entry: CustomEntry) => {
  const rom = roms[entry.game].rom;
  const dma = new DmaData(roms[entry.game].dma);
  const index = FILES_TO_INDEX[entry.game][entry.file];
  if (index === undefined) {
    throw new Error(`File ${entry.file} not found in game ${entry.game}`);
  }
  const dmaEntry = dma.read(index);
  return Buffer.from(rom.subarray(dmaEntry.virtStart, dmaEntry.virtEnd));
};

/* TODO: Cache this */
const makeSplitObject = async (roms: DecompressedRoms, entry: CustomEntry) => {
  const buf = await getObjectBuffer(roms, entry);
  const obj = splitObject(buf, entry.offsets);

  if (!process.env.ROLLUP) {
    const outDir = path.resolve('build', 'custom');
    const outBasename = entry.name.toLowerCase();
    const outFilename = path.resolve(outDir, `${outBasename}.zobj`);
    await fs.mkdir(outDir, { recursive: true });
    await fs.writeFile(outFilename, obj.data);
  }

  return obj;
};

export const custom = async (monitor: Monitor, roms: DecompressedRoms) => {
  monitor.log("Building custom objects");
  const cgPath = process.env.ROLLUP ? '' : path.resolve('build', 'include', 'combo', 'custom.h');
  const cg = new CodeGen(cgPath, 'CUSTOM_H');
  const objects = await Promise.all(ENTRIES.map(x => makeSplitObject(roms, x)));
  const objectDmaBuffer = Buffer.alloc(0x10 * objects.length);
  const objectDma = new DmaData(objectDmaBuffer);
  const objTable = Buffer.alloc(0x08 * objects.length);
  let vaddr = 0x08000000;
  let paddr = CUSTOM_ADDR;
  const data: Buffer[] = [];
  for (let i = 0; i < objects.length; ++i) {
    const obj = objects[i];
    const objEntry = ENTRIES[i];
    const objCompressed = await compressFile(obj.data);
    const objCompressedSize = align(objCompressed.byteLength, 0x10);
    const virtStart = vaddr;
    const virtEnd = vaddr + obj.data.byteLength;
    const physStart = paddr;
    const physEnd = paddr + objCompressedSize;
    objectDma.write(i, { virtStart, virtEnd, physStart, physEnd });
    objTable.writeUInt32BE(virtStart, i * 8 + 0);
    objTable.writeUInt32BE(virtEnd, i * 8 + 4);
    data.push(objCompressed);
    if (objCompressed.byteLength !== objCompressedSize) {
      data.push(Buffer.alloc(objCompressedSize - objCompressed.byteLength));
    }
    vaddr = virtEnd;
    paddr += objCompressedSize;
    cg.define(['CUSTOM_OBJECT_ID', objEntry.name].join('_'), 0x2000 | i);
    for (let j = 0; j < obj.offsets.length; ++j) {
      const offset = obj.offsets[j];
      cg.define(['CUSTOM_OBJECT', objEntry.name, j].join('_'), offset);
    }
  }
  cg.define('CUSTOM_DMA_ADDR', paddr);
  cg.define('CUSTOM_DMA_SIZE', objects.length);
  data.push(objectDmaBuffer);
  paddr += objectDmaBuffer.byteLength;
  cg.define('CUSTOM_OBJECTS_ADDR', paddr);
  cg.define('CUSTOM_OBJECTS_SIZE', objects.length);
  data.push(objTable);

  if (!process.env.ROLLUP) {
    await cg.emit();
  }

  return Buffer.concat(data);
};
